lib/IDE/Refactoring.cpp

```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
   return false;
 }
 
+class ExpandableTernaryExprInfo {
+
+public:
+  virtual ~ExpandableTernaryExprInfo() {}
+
+  virtual swift::IfExpr *getIf() = 0;
+
+  virtual swift::SourceRange getNameRange() = 0;
+
+  virtual swift::Type getType() = 0;
+
+  virtual bool shouldDeclareNameAndType() {
+    return !getType().isNull();
+  }
+
+  virtual bool isValid() {
+
+    //Ensure all public properties are non-nil and valid
+    if (!getIf() || !getNameRange().isValid())
+      return false;
+    if (shouldDeclareNameAndType() && getType().isNull())
+      return false;
+
+    return true; //valid
+  }
+
+  swift::CharSourceRange getNameCharRange(const swift::SourceManager &SM) {
+    return Lexer::getCharSourceRangeFromSourceRange(SM, getNameRange());
+  }
+};
+
+class ExpandableAssignTernaryExprInfo: public ExpandableTernaryExprInfo {
+
+public:
+  ExpandableAssignTernaryExprInfo(swift::AssignExpr *Assign): Assign(Assign) {}
+
+  swift::IfExpr *getIf() {
+    if (!Assign)
+      return nullptr;
+
+    return dyn_cast<IfExpr>(Assign->getSrc());
+  }
+
+  swift::SourceRange getNameRange() {
+    auto Invalid = swift::SourceRange();
+
+    if (!Assign)
+      return Invalid;
+
+    if (auto dest = Assign->getDest())
+      return dest->getSourceRange();
+
+    return Invalid;
+  }
+
+  swift::Type getType() {
+    return swift::Type();
+  }
+
+private:
+  swift::AssignExpr *Assign = nullptr;
+};
+
+class ExpandableBindingTernaryExprInfo: public ExpandableTernaryExprInfo {
+
+public:
+  ExpandableBindingTernaryExprInfo(swift::PatternBindingDecl *Binding):
+  Binding(Binding) {}
+
+  swift::IfExpr *getIf() {
+    if (Binding && Binding->getNumPatternEntries() == 1)
+      return dyn_cast<IfExpr>(Binding->getInit(0));
+
+    return nullptr;
+  }
+
+  swift::SourceRange getNameRange() {
+    if (auto Pattern = getNamePattern())
+      return Pattern->getSourceRange();
+
+    return swift::SourceRange();
+  }
+
+  swift::Type getType() {
+    if (auto Pattern = getNamePattern())
+      return Pattern->getType();
+
+    return swift::Type();
+  }
+
+private:
+  swift::Pattern *getNamePattern() {
+    if (!Binding || Binding->getNumPatternEntries() != 1)
+      return nullptr;
+
+    auto Pattern = Binding->getPattern(0);
+
+    if (!Pattern)
+      return nullptr;
+
+    if (auto TyPattern = dyn_cast<TypedPattern>(Pattern))
+      Pattern = TyPattern->getSubPattern();
+
+    return Pattern;
+  }
+
+  swift::PatternBindingDecl *Binding = nullptr;
+};
+
+std::unique_ptr<ExpandableTernaryExprInfo>
+findExpandableTernaryExpression(ResolvedRangeInfo Info) {
+
+  if (Info.Kind != RangeKind::SingleDecl
+      && Info.Kind != RangeKind:: SingleExpression)
+    return nullptr;
+
+  if (Info.ContainedNodes.size() != 1)
+    return nullptr;
+
+  if (auto D = Info.ContainedNodes[0].dyn_cast<Decl*>()) {
+    auto Binding = dyn_cast<PatternBindingDecl>(D);
```

Comment:
Check if `Binding` is nullptr here and only create `ExpandableBindingTernaryExprInfo` when it is not. Therefore, we don't need to check whether `Binding` is null inside `ExpandableBindingTernaryExprInfo`.